recipes-core/searcher-ssh-key/files/mode-controller.c

```diff
@@ -0,0 +1,391 @@
+    if (!f) return -1;
+
+    fprintf(f, "%s\n", mode);
+    fclose(f);
+    return 0;
+}
+
+static int kill_existing_connections(void) {
+    uid_t real_uid = getuid();
+    uid_t real_gid = getgid();
+    int ret = 0;
+    FILE *fp;
+    char line[256];
+    char *endptr;
+    long pid;
+
+    // Elevate privileges
+    if (setuid(0) != 0 || setgid(0) != 0) {
+        log_message(LOG_ERR, "Failed to elevate privileges for killing connections");
+        return -1;
+    }
+
+    // Using BusyBox ps to get PIDs
+    fp = popen("ps | grep ^'[ ]*[0-9]'", "r");
+    if (!fp) {
+        log_message(LOG_ERR, "Failed to get process list");
+        if (setgid(real_gid) != 0 || setuid(real_uid) != 0) {
+            log_message(LOG_ERR, "Failed to drop privileges after popen error");
+        }
+        return -1;
+    }
+
+    // Skip header line
+    fgets(line, sizeof(line), fp);
+
+    // Process each line
+    while (fgets(line, sizeof(line), fp)) {
+        pid = strtol(line, &endptr, 10);
+        if (pid > 1) {  // Skip init process
+            struct stat st;
+            char procpath[32];
+            snprintf(procpath, sizeof(procpath), "/proc/%ld", pid);
+            
+            if (stat(procpath, &st) == 0 && st.st_uid == SEARCHER_UID) {
+                log_message(LOG_INFO, "Sending SIGTERM to process %ld", pid);
+                if (kill(pid, SIGTERM) != 0) {
+                    log_message(LOG_WARNING, "Failed to send SIGTERM to PID %ld: %s", 
+                              pid, strerror(errno));
+                }
+            }
+        }
+    }
+    pclose(fp);
+
+    // Wait for processes to terminate
+    sleep(2);
+
+    // Force kill remaining processes
+    fp = popen("ps | grep ^'[ ]*[0-9]'", "r");
+    if (!fp) {
+        if (setgid(real_gid) != 0 || setuid(real_uid) != 0) {
+            log_message(LOG_ERR, "Failed to drop privileges after second popen error");
+        }
+        return -1;
+    }
+
+    // Skip header line
+    fgets(line, sizeof(line), fp);
+
+    while (fgets(line, sizeof(line), fp)) {
+        pid = strtol(line, &endptr, 10);
+        if (pid > 1) {
+            struct stat st;
+            char procpath[32];
+            snprintf(procpath, sizeof(procpath), "/proc/%ld", pid);
+            
+            if (stat(procpath, &st) == 0 && st.st_uid == SEARCHER_UID) {
+                log_message(LOG_INFO, "Sending SIGKILL to process %ld", pid);
+                if (kill(pid, SIGKILL) != 0) {
+                    log_message(LOG_WARNING, "Failed to send SIGKILL to PID %ld: %s", 
+                              pid, strerror(errno));
+                    ret = -1;
+                }
+            }
+        }
+    }
+    pclose(fp);
+
+    // Drop privileges
+    if (setgid(real_gid) != 0 || setuid(real_uid) != 0) {
+        log_message(LOG_ERR, "Failed to drop privileges");
+        return -1;
+    }
+
+    return ret;
+}
+
+static int execute_transition_script(const char *from_mode, const char *to_mode) {
```

Comment:
I'd guess the function would be simpler if you hardcoded the transitions as one function each